lens/registry.go

```diff
@@ -0,0 +1,339 @@
+	"github.com/sourcenetwork/defradb/core"
+	"github.com/sourcenetwork/defradb/datastore"
+)
+
+// todo: This file, particularly the `lensLocker`` stuff, contains fairly sensitive code that is both
+// cumbersome to fully test with integration/benchmark tests, and can have a significant affect on
+// the users if broken (deadlocks, large performance degradation).  It should have proper unit tests.
+// https://github.com/sourcenetwork/defradb/issues/1596
+
+// lensRegistry is responsible for managing all migration related state within a local
+// database instance.
+type lensRegistry struct {
+	lensPoolSize int
+	// lens lockers by source schema version ID
+	lensWarehouse map[string]*lensLocker
+	// lens configurations by source schema version ID
+	lensConfigs map[string]client.LensConfig
+}
+
+var _ client.LensRegistry = (*lensRegistry)(nil)
+
+// DefaultPoolSize is the default size of the lens pool for each schema version.
+const DefaultPoolSize int = 5
+
+// NewRegistry instantiates a new registery.
+//
+// It will be of size 5 (per schema version) if a size is not provided.
+func NewRegistry(lensPoolSize immutable.Option[int]) *lensRegistry {
+	var size int
+	if lensPoolSize.HasValue() {
+		size = lensPoolSize.Value()
+	} else {
+		size = DefaultPoolSize
+	}
+
+	return &lensRegistry{
+		lensPoolSize:  size,
+		lensWarehouse: map[string]*lensLocker{},
+		lensConfigs:   map[string]client.LensConfig{},
+	}
+}
+
+func (r *lensRegistry) SetMigration(ctx context.Context, txn datastore.Txn, cfg client.LensConfig) error {
+	key := core.NewSchemaVersionMigrationKey(cfg.SourceSchemaVersionID)
+
+	json, err := json.Marshal(cfg)
+	if err != nil {
+		return err
+	}
+
+	err = txn.Systemstore().Put(ctx, key.ToDS(), json)
+	if err != nil {
+		return err
+	}
+
+	err = r.cacheLens(txn, cfg)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *lensRegistry) cacheLens(txn datastore.Txn, cfg client.LensConfig) error {
+	locker, lockerAlreadyExists := r.lensWarehouse[cfg.SourceSchemaVersionID]
+	if !lockerAlreadyExists {
+		locker = newLocker(r.lensPoolSize, cfg)
+	}
+
+	newLensPipes := make([]*lensPipe, r.lensPoolSize)
+	for i := 0; i < r.lensPoolSize; i++ {
+		var err error
+		newLensPipes[i], err = newLensPipe(cfg)
+		if err != nil {
+			return err
+		}
+	}
+
+	// todo - handling txns like this means that the migrations are not available within the current
+	// transaction if used for stuff (e.g. GQL requests) before commit.
+	// https://github.com/sourcenetwork/defradb/issues/1592
+	txn.OnSuccess(func() {
+		if !lockerAlreadyExists {
+			r.lensWarehouse[cfg.SourceSchemaVersionID] = locker
+		}
+
+	drainLoop:
+		for {
+			select {
+			case <-locker.safes:
+			default:
+				break drainLoop
+			}
+		}
+
+		for _, lensPipe := range newLensPipes {
+			locker.returnLens(lensPipe)
+		}
+
+		r.lensConfigs[cfg.SourceSchemaVersionID] = cfg
+	})
+
+	return nil
+}
+
+func (r *lensRegistry) ReloadLenses(ctx context.Context, txn datastore.Txn) error {
```

Comment:
todo: there are no tests that cover `ReloadLenses`

Reply:
There are, and this is noted in a PR comment.